db_stress_tool/myrocks_style_txns_stress.cc

```diff
@@ -0,0 +1,1275 @@
+  uint32_t rand = thread->rand.Uniform(3);
+  Status s;
+  if (0 == rand) {
+    // Update primary key.
+    uint32_t old_a = ChooseA(thread);
+    uint32_t new_a = GenerateNextA();
+    s = PrimaryKeyUpdateTxn(thread, old_a, new_a);
+  } else if (1 == rand) {
+    // Update secondary key.
+    uint32_t old_c = thread->rand.Next() % kInitNumC;
+    int count = 0;
+    uint32_t new_c = 0;
+    do {
+      ++count;
+      new_c = thread->rand.Next() % kInitNumC;
+    } while (count < 100 && new_c == old_c);
+    if (count >= 100) {
+      // If we reach here, it means our random number generator has a serious
+      // problem, or kInitNumC is chosen poorly.
+      std::terminate();
+    }
+    s = SecondaryKeyUpdateTxn(thread, old_c, new_c);
+  } else if (2 == rand) {
+    // Update primary index value.
+    uint32_t a = ChooseA(thread);
+    s = UpdatePrimaryIndexValueTxn(thread, a, /*b_delta=*/1);
+  } else {
+    // Should never reach here.
+    assert(false);
+  }
+  return s;
+}
+
+Status MyRocksStyleTxnsStressTest::PrimaryKeyUpdateTxn(ThreadState* thread,
+                                                       uint32_t old_a,
+                                                       uint32_t new_a) {
+#ifdef ROCKSDB_LITE
+  (void)thread;
+  (void)old_a;
+  (void)new_a;
+  return Status::NotSupported();
+#else
+  std::string old_pk = MyRocksRecord::EncodePrimaryKey(old_a);
+  std::string new_pk = MyRocksRecord::EncodePrimaryKey(new_a);
+  Transaction* txn = nullptr;
+  WriteOptions wopts;
+  Status s = NewTxn(wopts, &txn);
+  if (!s.ok()) {
+    assert(!txn);
+    thread->stats.AddErrors(1);
+    return s;
+  }
+  txn->SetSnapshotOnNextOperation(/*notifier=*/nullptr);
+
+  const Defer cleanup([&s, thread, txn, this]() {
+    if (s.ok()) {
+      // Two gets, one for existing pk, one for locking potential new pk.
+      thread->stats.AddGets(/*ngets=*/2, /*nfounds=*/1);
+      thread->stats.AddDeletes(1);
+      thread->stats.AddBytesForWrites(
+          /*nwrites=*/2, MyRocksRecord::kPrimaryIndexEntrySize +
+                             MyRocksRecord::kSecondaryIndexEntrySize);
+      thread->stats.AddSingleDeletes(1);
+      return;
+    }
+    if (s.IsNotFound()) {
+      thread->stats.AddGets(/*ngets=*/1, /*nfounds=*/0);
+    } else if (s.IsBusy()) {
+      // ignore.
+    } else {
+      thread->stats.AddErrors(1);
+    }
+    RollbackTxn(txn).PermitUncheckedError();
+  });
+
+  ReadOptions ropts;
+  std::string value;
+  s = txn->GetForUpdate(ropts, old_pk, &value);
+  if (!s.ok()) {
+    return s;
+  }
+  std::string empty_value;
+  s = txn->GetForUpdate(ropts, new_pk, &empty_value);
+  if (s.ok()) {
+    s = Status::Busy();
+    return s;
+  }
+
+  auto result = MyRocksRecord::DecodePrimaryIndexValue(value);
+  s = std::get<0>(result);
+  if (!s.ok()) {
+    return s;
+  }
+  uint32_t b = std::get<1>(result);
+  uint32_t c = std::get<2>(result);
+
+  s = txn->Delete(old_pk);
```

Comment:
In general, we call Put/Delete with `assume_tracked` on, on the assumption that we always read+lock before modifying (ie. no blind writes).